### PR TITLE
Add TimeZoneFallbackOverride date field length; use NeoTimeZoneSkeleton

### PR DESCRIPTION
**Pull request by sffc**
#1317

### CODE REVIEW

File: components/datetime/src/neo_zone.rs [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+/// A type that has a time zone configuration associated with it.
+// TODO: Make this require sealed
+pub trait HasZoneComponents {
+    /// The configuration for formatting this time zone.
+    const ZONE_CONFIG: NeoZoneComponents;
+}
+
+/// "GMT-8"
+#[derive(Debug)]
+pub enum GmtShort {}
```
Comment (Manishearth) [inline components/datetime/src/neo_zone.rs:118]: thought: not sure if empty enums are what we need for these markers but that's a separate discussion to be had later

**Review by Manishearth — COMMENTED** [review]
I'm not sure what preference I have here. The private use fields are a bit messy but not terrible. This also seems _fine_?

Interested in what Zibi thinks

**Comment by sffc** [comment]
I'm leaning toward private use field lengths unless I hear otherwise. All the code is just so intertwined with PatternItems (for good reason), and there's not a good reason to break from that.

**Comment by sffc** [comment]
Okay I've tried implementing option 1. Some more thoughts:

- Option 1 requires three ways to represent mostly the same thing: a ZoneComponents enum of all supported configurations (primary public interface); a Field used to represent that enum in a pattern string (public but low level); and the ZoneConfig as the more useful machine representation (mostly internal but it needs to be public because it is in a trait). In comparison, Option 2 can get away with only ZoneConfig and not the others.
- It is nice that Option 1 works seamlessly with existing pattern strings. Option 2 requires some pre-processing at some point to handle pattern strings that might contain CLDR time zone items.

**Comment by zbraniecki** [comment]
I don't have a strong opinion, I'm sorry. Can we do the Option 2 pre-processing at data build time?

**Comment by sffc** [comment]
Ok this has been rewritten multiple times but I simplified the change to only use the private use field length and it's now ready for review.

**Comment by sffc** [comment]
Backup of neo_zone.rs:

```rust
// This file is part of ICU4X. For terms of use, please see the file
// called LICENSE at the top level of the ICU4X source tree
// (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).

//! Types for neo time zone formatting.

use crate::fields::{self, Field, FieldSymbol, FieldLength};
#[cfg(feature = "experimental")]
use crate::neo_marker::private;
use crate::time_zone::Iso8601Format;

/// Time zone offset formatting style.
#[derive(Debug, Copy, Clone)]
pub enum NeoOffsetStyle {
    /// "GMT-8"
    /// `O`
    GmtShort,
    /// "GMT-08:00"
    /// `OOOO`, `ZZZZ`
    GmtLong,
    /// All `x` and `X` formats
    Iso8601(Iso8601Format),
}

/// Time zone formatting style, not including offsets.
#[derive(Debug, Copy, Clone)]
pub enum NeoZoneStyle {
    /// "PDT"
    /// `z..zzz`
    SpecificShort,
    /// "Pacific Daylight Time"
    /// `zzzz`
    SpecificLong,
    /// "PT"
    /// `v`
    GenericShort,
    /// "Pacific Time"
    /// `vvvv`
    GenericLong,
    /// "uslax"
    /// `V`
    IdentifierShort,
    /// "America/Los_Angeles"
    /// `VV`
    IdentifierLong,
    /// "Los Angeles"
    /// `VVV`
    City,
    /// "Los Angeles Time"
    /// `VVVV`
    GenericLocation,
    /// "Pacific Time (Los Angeles)"
    /// not available via field?
    GenericPartialLocation,
}

/// A specification for the desired display of a time zone.
///
/// Unlike date and time, we support only a single time zone component, but the
/// specific component can change.
///
/// Since the length of a time zone can vary independent of the date and time,
/// the time zone lengths are directly encoded into this enum.
#[derive(Debug, Copy, Clone)]
#[non_exhaustive]
pub struct NeoZoneConfig {
    /// The primary time zone style.
    pub first: Option<NeoZoneStyle>,
    /// The fallback time zone style if the primary style is unavailable.
    pub second: Option<NeoZoneStyle>,
    /// The offset style if the time zone cannot be formatted.
    pub offset: NeoOffsetStyle,
}

impl NeoZoneConfig {
    pub(crate) fn from_field(
        field: fields::TimeZone,
        length: fields::FieldLength,
    ) -> Option<Self> {
        match (field, length) {
            // TODO: This should actually be GenericShortOrLocation according to spec
            (fields::TimeZone::LowerV, fields::FieldLength::One) => Some(GenericShortOrGmt::ZONE_CONFIG),
            _ => todo!()
        }
    }

    pub(crate) fn default_for_fallback() -> Self {
        Self {
            first: None,
            second: None,
            offset: NeoOffsetStyle::Iso8601(Iso8601Format::default_for_fallback()),
        }
    }
}

/// A type that has a time zone configuration associated with it.
// TODO: Make this require sealed
pub trait HasZoneComponents {
    /// The configuration for formatting this time zone.
    const ZONE_CONFIG: NeoZoneConfig;
    /// The corresponding field.
    const ZONE_FIELD: Field;
}

/// "GMT-8"
#[derive(Debug)]
pub enum GmtShort {}

#[cfg(feature = "experimental")]
impl private::Sealed for GmtShort {}

impl HasZoneComponents for GmtShort {
    const ZONE_CONFIG: NeoZoneConfig = NeoZoneConfig {
        first: None,
        second: None,
        offset: NeoOffsetStyle::GmtShort,
    };
    const ZONE_FIELD: Field = Field {
        symbol: FieldSymbol::TimeZone(fields::TimeZone::UpperO),
        length: FieldLength::One,
    };
}

/// "PT" or "GMT-8"
#[derive(Debug)]
pub enum GenericShortOrGmt {}

#[cfg(feature = "experimental")]
impl private::Sealed for GenericShortOrGmt {}

impl HasZoneComponents for GenericShortOrGmt {
    const ZONE_CONFIG: NeoZoneConfig = NeoZoneConfig {
        first: Some(NeoZoneStyle::GenericShort),
        second: None,
        offset: NeoOffsetStyle::GmtShort,
    };
    const ZONE_FIELD: Field = Field {
        symbol: FieldSymbol::TimeZone(fields::TimeZone::LowerV),
        length: FieldLength::TimeZoneFallbackOverride(fields::TimeZoneFallbackOverride::ShortOrGmt),
    };
}

/// "Los Angeles Time" or "GMT-8"
#[derive(Debug)]
pub enum GenericLocationOrGmt {}

#[cfg(feature = "experimental")]
impl private::Sealed for GenericLocationOrGmt {}

impl HasZoneComponents for GenericLocationOrGmt {
    const ZONE_CONFIG: NeoZoneConfig = NeoZoneConfig {
        first: Some(NeoZoneStyle::GenericLocation),
        second: None,
        offset: NeoOffsetStyle::GmtShort,
    };
    const ZONE_FIELD: Field = Field {
        symbol: FieldSymbol::TimeZone(fields::TimeZone::UpperV),
        length: FieldLength::Wide,
    };
}

/// "PT" or "Los Angeles Time" or "GMT-8"
#[derive(Debug)]
pub enum GenericShortOrLocation {}

#[cfg(feature = "experimental")]
impl private::Sealed for GenericShortOrLocation {}

impl HasZoneComponents for GenericShortOrLocation {
    const ZONE_CONFIG: NeoZoneConfig = NeoZoneConfig {
        first: Some(NeoZoneStyle::GenericShort),
        second: Some(NeoZoneStyle::GenericLocation),
        offset: NeoOffsetStyle::GmtShort,
    };
    const ZONE_FIELD: Field = Field {
        symbol: FieldSymbol::TimeZone(fields::TimeZone::LowerV),
        length: FieldLength::One,
    };
}
```

**Review by Manishearth — APPROVED** [review]

**Pull request merged**